src/lib.rs

```diff
@@ -607,48 +608,196 @@ where
     /// Copying four bytes within a slice:
     ///
     /// ```
-    /// use volatile::Volatile;
+    /// extern crate core;
+    /// use volatile::VolatilePtr;
+    /// use core::ptr::NonNull;
     ///
     /// let mut byte_array = *b"Hello, World!";
     /// let mut slice: &mut [u8] = &mut byte_array[..];
-    /// let mut volatile = Volatile::new(slice);
+    /// let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(slice)) };
     ///
     /// volatile.copy_within(1..5, 8);
     ///
     /// assert_eq!(&byte_array, b"Hello, Wello!");
-    #[cfg(feature = "unstable")]
     pub fn copy_within(&mut self, src: impl RangeBounds<usize>, dest: usize)
     where
         T: Copy,
-        R: DerefMut,
     {
+        let len = self.pointer.len();
         // implementation taken from https://github.com/rust-lang/rust/blob/683d1bcd405727fcc9209f64845bd3b9104878b8/library/core/src/slice/mod.rs#L2726-L2738
         let Range {
             start: src_start,
             end: src_end,
-        } = range(src, ..self.reference.len());
+        } = range(src, ..len);
         let count = src_end - src_start;
-        assert!(
-            dest <= self.reference.len() - count,
-            "dest is out of bounds"
-        );
+        assert!(dest <= len - count, "dest is out of bounds");
         // SAFETY: the conditions for `volatile_copy_memory` have all been checked above,
         // as have those for `ptr::add`.
         unsafe {
             intrinsics::volatile_copy_memory(
-                self.reference.as_mut_ptr().add(dest),
-                self.reference.as_ptr().add(src_start),
+                self.pointer.as_mut_ptr().add(dest),
+                self.pointer.as_mut_ptr().add(src_start),
                 count,
             );
         }
     }
+
+    pub fn split_at(
+        &self,
+        mid: usize,
+    ) -> (
+        VolatilePtr<[T], Access<R, access::NoAccess>>,
+        VolatilePtr<[T], Access<R, access::NoAccess>>,
+    ) {
+        assert!(mid <= self.pointer.len());
+        // SAFETY: `[ptr; mid]` and `[mid; len]` are inside `self`, which
+        // fulfills the requirements of `from_raw_parts_mut`.
+        unsafe { self.split_at_unchecked(mid) }
+    }
+
+    pub fn split_at_mut(
+        &mut self,
+        mid: usize,
+    ) -> (
+        VolatilePtr<[T], Access<R, W>>,
+        VolatilePtr<[T], Access<R, W>>,
+    ) {
```

Comment:
Could this method take `Self` as an owned value? A problem with the current implementation is that the returned `VolatilePtr`s can't outlive the reference to the `VolatilePtr`, but they really only need to not outlive the lifetime on the `VolatilePtr`. 

This leads to problems in practice:
```rust
fn fetch_more_data() -> [u8; 4] {
    todo!("fetch some data")
}

fn read_some_data(mut dest: VolatilePtr<[u8], Access<SafeAccess, SafeAccess>>) {
    while dest.len() != 0 {
        let chunk = fetch_more_data();

        let len = cmp::min(chunk.len(), dest.len());
        let (mut head, tail) = dest.split_at_mut(len); // <-- `dest` does not live long enough
        head.copy_from_slice(&chunk[..len]);
        dest = tail; // <-- cannot assign to `dest` because it is borrowed
    }
}
```
(The same code works just fine with `&mut [u8]`.)

Reply:
I've been trying to prototype an updated implementation for this, and I've also realized that having _all_ the methods take `self` is probably what we want to do.

Conceptually, it makes sense for our `VolatilePtr` have `self` methods as it's a `Copy` type. It also reduces the overall number of `mut` vs not-`mut` method variants.